### PR TITLE
stylo: Add a build log mechanism.

### PR DESCRIPTION
**Pull request by emilio**
This will allow us to debug bindgen problems.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16022)

### CODE REVIEW

File: components/style/build_gecko.rs [hunk 1]
```diff
@@ -517,6 +517,49 @@ mod bindings {
...
+                self.file.is_some() && meta.target().contains(&self.filter)
+            }
+
+            fn log(&self, record: &log::LogRecord) {
+                if !self.enabled(record.metadata()) {
+                    return;
+                }
+
+                let mut file = self.file.as_ref().unwrap().lock().unwrap();
+                let _ =
```
Comment (upsuper) [inline components/style/build_gecko.rs]: Why do you ignore the result? Would it be better to just use `unwrap()`?

File: components/style/build_gecko.rs [hunk 2]
```diff
@@ -517,6 +517,49 @@ mod bindings {
...
+                             record.location().file(),
+                             record.location().line());
+            }
+        }
+
+        log::set_logger(|log_level| {
+            log_level.set(log::LogLevelFilter::Debug);
+            Box::new(BuildLogger {
+                file: env::var("STYLO_BUILD_LOG").ok().and_then(|path| {
+                    fs::File::create(path).ok().map(Mutex::new)
```
Comment (upsuper) [inline components/style/build_gecko.rs]: Where would the base be if a relative path is specified?

**Review by upsuper — APPROVED** [review]

File: components/style/build_gecko.rs (hunk 2 above)
Comment (emilio) [inline components/style/build_gecko.rs]: Relative to the style crate. I can change it if needed, but seemed like unnecessary complication. I've used it with `STYLO_BUILD_LOG=/tmp/build.log` and similar.

File: components/style/build_gecko.rs (hunk 1 above)
Comment (emilio) [inline components/style/build_gecko.rs]: I have no real preference, I thought that we shouldn't panic just for failing to log to disk, but...

Comment (upsuper) [inline components/style/build_gecko.rs]: Yeah, that proably makes sense.

File: components/style/build_gecko.rs (hunk 2 above)
Comment (emilio) [inline components/style/build_gecko.rs]: In particular, the ideal would be to find the user's PWD, but that's not exposed as far as I can tell. The workspace root in Gecko isn't also anywhere you'd specially expect your logs to appear, and I guess I can always give another env var to the build script, but seemed unnecessary.

**Comment by emilio** [comment]
@bors-servo r=upsuper

 * Let me know if you want me to change the relative path, happy to do so in a followup.

File: components/style/build_gecko.rs (hunk 2 above)
Comment (upsuper) [inline components/style/build_gecko.rs]: Probably fine, as far as we just explicitly specify an absolute path to it in Gecko's build system.

Comment (emilio) [inline components/style/build_gecko.rs]: I don't think we want to enable this by default for everyone, just when problems arise. This makes builds so much slower.

Comment (emilio) [inline components/style/build_gecko.rs]: (And yeah, I know I initially thought otherwise, but after trying that I changed my mind :P)

**Pull request merged**